src/main/java/net/minecraftforge/client/gui/tabs/GuiTab.java

```diff
@@ -0,0 +1,318 @@
+        GlStateManager.enableBlend(); // Make sure blend is enabled else tabs show a white border.
+        guiContainer.drawTexturedModalRect(x, y, texX, texY, 28, 32);
+        x += 6;
+        y += 8 + (isFirstRow ? 1 : -1);
+        GlStateManager.enableLighting();
+        if (!getIconStack().isEmpty())
+        {
+            ItemStack itemstack = getIconStack();
+            itemRender.renderItemAndEffectIntoGUI(itemstack, x, y);
+            itemRender.renderItemOverlays(fontRenderer, itemstack, x, y);
+        }
+        else if (getIconResLoc() != null)
+        {
+            Minecraft.getMinecraft().getTextureManager().bindTexture(getIconResLoc());
+            Gui.drawModalRectWithCustomSizedTexture(x, y, 0f, 0f, 16, 16, 16f, 16f);
+        }
+        GlStateManager.disableLighting();
+    }
+
+    public boolean isAlignedRight()
+    {
+        return this.getColumn() == 5;
+    }
+
+    /**
+     * Returns the index relative to the gui this tab was registered too
+     */
+    public int getIndex()
+    {
+        return tabRegistry.get(parentContainer).indexOf(this);
+    }
+
+    /**
+     * Whether this tab is on the top row or not
+     */
+    public boolean isTopRow()
+    {
+        return (getIndex() > 11 ? (getIndex() - 12) % 12 : this.getIndex()) < 6;
+    }
+
+    /**
+     * Returns index % 6 to determine column
+     */
+    public int getColumn()
+    {
+        return (getIndex() > 11 ? (getIndex() - 12) % 12 : this.getIndex()) % 6;
+    }
+
+    /**
+     * @return Whether this tab's parent GUI is the current GUI
+     */
+    public boolean isActiveTab(Class<? extends GuiContainer> guiContainer)
+    {
+        return guiContainer == getTargetGui();
+    }
+
+    /**
+     * @return The tab index / 12
+     */
+    public int getPage()
+    {
+        return getIndex() / 12;
+    }
+
+    /**
+     * @return The unlocalized name for this tab
+     */
+    public String getUnlocalizedName()
+    {
+        return "guiGroup." + name;
+    }
+
+    public ResourceLocation getIconResLoc()
+    {
+        return iconResLoc;
+    }
+
+    public ItemStack getIconStack()
+    {
+        return iconStack;
+    }
+
+    /**
+     * @return A list of the appended tabs for the given GuiContainer
+     */
+    public static List<GuiTab> getTabsForGui(Class<? extends GuiContainer> guiContainer)
+    {
+        return tabRegistry.get(guiContainer);
+    }
+
+    /**
+     * Adds the GuiTab to GuiContainer
+     *
+     * @param parentContainer The GuiContainer to add a tab too
+     * @return The GuiTab
+     */
+    public GuiTab addTo(Class<? extends GuiContainer> parentContainer)
+    {
+        if (parentContainer == GuiContainerCreative.class)
+        {
+            throw new IllegalArgumentException("Cannot add tabs to creative inventory. Use CreativeTabs instead");
+        }
+        else if (parentContainer == InventoryEffectRenderer.class)
```

Comment:
Is it worth checking for all superclasses? Currently this could be bypassed by adding to GuiContainer.class

Reply:
What exactly do you suggest ?

Reply:
Class::isSuperClass or something along those lines (can't remember the exact method)

Reply:
Sorry, I ment like what do you mean by
> Is it worth checking for all superclasses? Currently this could be bypassed by adding to GuiContainer.class

Reply:
You have put a check to prevent adding to creative inventory, even when a superclass is used. But you are manually specifying the classes to disallow instead of using the method that checks for all superclasses

Reply:
so like:
```java
        if (parentContainer == GuiContainerCreative.class)
        {
            throw new IllegalArgumentException("Cannot add tabs to creative inventory. Use CreativeTabs instead");
        }
        else if (parentContainer == GuiContainerCreative.class.getSuperclass())
        {
            throw new IllegalArgumentException(name + " tab has a parent that is ambiguous with the creative inventory. cannot proceed with registry.");
        }
```
?

Reply:
No, that is just checking that parentClass is InventoryEffectRenderer's superclass.
I think you need to use `parentClass.isAssignableFrom(InventoryEffectRenderer.class)` or something along those lines. Look into the methods of Class and you will find one which can check for sub/super classes

Reply:
Hmm, going to test it in a few... I've got some thoughts that this might block some of the vanilla classeses from functioning with tabs like the PlayerInventory container, which extends InventoryEffectRenderer.. I might be wrong (which is more likely)

Reply:
As long as it is a subclass of InventoryEffectRenderer it won't be affected, only InventoryEffectRenderer and it's superclasses

Reply:
Alrighty, everything works as expected.. thanks for the suggestion!